### DISCUSSION TITLE
React Toolkit with vite-plugin-ssr

### DISCUSSION THREAD

**Discussion by arp** [vikejs/vike#1021] [Discussion] [answered]
I've built an app which uses RTK on top of vite-plugin-ssr. I can't figure out how to make it run without library import problems both in dev and production.

This works in development:

```
import { createSlice, createSelector, PayloadAction } from "@reduxjs/toolkit"
```

it builds correctly, but then fails in `vite preview`:

```
import { createSlice as createSlice$1, createSelector } from "@reduxjs/toolkit";
                                       ^^^^^^^^^^^^^^
SyntaxError: Named export 'createSelector' not found. The requested module '@reduxjs/toolkit' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@reduxjs/toolkit';
const { createSlice: createSlice$1, createSelector } = pkg;
```

When I follow the suggested hint, `vite preview` starts to work correctly:

```
  import rtkPkg from "@reduxjs/toolkit/dist"
  const { createSlice, createSelector, PayloadAction } = rtkPkg
```

(using "dist" there is a suggestion I found in RTK discussions as they say not all of their modules are ESM, some are CommonJS)

but then it fails in `vite dev`, printing this in the browser's debug console:

```
Uncaught (in promise) TypeError: rtkQueryPkg is undefined
```

It looks as if `dev` and `build` run through a different set of transforms and something just isn't imported the same way. Is there a recommended way to achieve consistency between dev and prod import behavior?

**Comment by brillout** [comment]
Can you create a reproduction? Smoothing out these issues is on the roadmap, and the more reproductions I can get the better.

To answer your question: https://github.com/cyco130/vite-plugin-cjs-interop but there may be a simpler solution which I'll try one I get my hands on a reproduction.

**Comment by arp** [comment, marked as answer]
Here's an example of how I was able to fix it for me:

```
  const rtkQueryPkg = require('@reduxjs/toolkit/dist/query/react')
  const { fetchBaseQuery, FetchBaseQueryArgs, FetchBaseQueryResult } = rtkQueryPkg
```

this works both in dev and production. From what I understand, RTK team is working on 2.0.0 version which will not have mixed imports, so the problem is going to be eventually solved: https://github.com/reduxjs/redux-toolkit/issues/1960. I'm not sure whether it's specific to Vite/VPS, it seems that other build pipelines are also affected.

**Reply by brillout** [reply]
Thanks for reporting back 👍.
